### PR TITLE
Allow reading ASCII table with duplicate column names

### LINKED ISSUE #8471: Table read fails with duplicated column names (with CSV format)
Trying to read in a CSV where there are multiple columns with the same name results in a rather cryptic ValueError:
``ValueError: Arguments "names" and "dtype" must match number of columns``

Opting out the fast reader it becomes a bit clear:
``ValueError: Duplicate column names``

In the zeroth order it would be great to receive the second message even in the first case. Then we may want to think about the pro and con of adding a new kwarg `rename_duplicate` here, too (we already have it for `add_column()`.

```
>>> from astropy.table import Table 
>>> input_text ='angDist,ra,dec,JNAME,ra,dec,errHalfMaj,errHalfMin,errPosAng,W1mag,W2mag,W3mag,W4mag,Jmag,Hmag,Kmag,e_W1mag,e_W2mag,e_W3mag,e_W4mag,e_Jmag,e_Hmag,e_Kmag,cntr,cc_flags,ext_flg,var_flg,ph_qual,r_2mass\n0.167960,2.11044940645274,0.688137082702486,J000826.51+004117.4,2.110478,+0.688174,0.150,0.143,142,14.232,14.199,12.449,8.773,15.150,14.637,14.347,0.031,0.056,,,0.043,0.052,0.064,15100001241034545,0000,0,10nn,AAUU,0.256\n0.537585,2.08882950663451,0.719216362471814,J000821.29+004308.8,2.088721,+0.719113,0.163,0.152,118,14.432,14.408,12.500,8.595,15.042,14.683,14.481,0.033,0.064,,,0.037,0.045,0.062,15100001241034551,0000,0,00nn,AAUU,0.534\n'

>>> Table.read(input_text, format='csv')
ValueError: Arguments "names" and "dtype" must match number of columns

>>> Table.read(input_text, format='csv', fast_reader=False)
ValueError: Duplicate column names
```

This example is coming from using the remote XMatch service with astroquery. I suppose as a workaround we can switch to use votable rather than csv (as the votable reader renames the columns automatically), but nevertheless it would be nice to fix this.

### PR DESCRIPTION
**Pull request by taldcroft**
### Description

This pull request is to allow reading ASCII tables with duplicate column names.  The strategy is the same s for adding a duplicate column name to a Table, namely adding `_<N>` to a duplicate name for increasing `N` until it is no longer a duplicate.

Ideally one might want to issue a warning in the case that duplicate column names were munged, and the first commit of this PR does that.  Unfortunately one runs into a situation that I can't figure out how to resolve: while guessing the file format one might end up generating a spurious warning.  This specifically happened in testing for a CDS file with a sentence at the top that had two instances of the word `of`.  During guessing this first line was interpreted as duplicate column names and emitted the warning.  So we have a choice between:

1. No warnings for duplicate column names
2. Possible spurious warnings for duplicates in some situations, in particular reading CDS files

The 2nd commit removed warnings, and that would be my preference, but this can go either way.

Fixes #8471

### CODE REVIEW

**Comment by hamogu** [comment]
3) As a third option, instead of emitting the warning, it could simply be written to a header attribute (not sure about order of things, e.g. does the header.meta dict exist at the point already)?

Then 3)a) would be: We don't output the warning, but it's available for debugging, similar to the guess list.
or 3)b) The warning is emitted as the last line in the read function. That's not as much a problem when guessing, because most guesses never get to the that line because they trigger some error before that.
or 3)c) That warning is emitted, but the guess logic catches the warnings, printing only the very last one to the screen.

That said, just because it's possible, I don't think we should do it. I'm fine with keeping it silent - printing the table will show that name mangling happened.

File: CHANGES.rst [hunk 1]
```diff
@@ -27,6 +27,9 @@ astropy.extern
 astropy.io.ascii
 ^^^^^^^^^^^^^^^^
 
+- Allow reading in ASCII tables which have duplicate column names.  In this
+  case a warning is issued and the names are adjusted to be unique. [#9939]
```
Comment (hamogu) [inline CHANGES.rst]: No warning is issued after commit 2 of this branch.

File: astropy/io/ascii/tests/test_read.py [hunk 2]
```diff
@@ -1493,3 +1495,43 @@ def test_kwargs_dict_guess(enable):
...
+    dat = ascii.read(lines, fast_reader=fast_reader)
+    assert dat.colnames == ['a', 'a_2', 'a_1', 'a_3', 'a_4']
+
+    with ExitStack() as stack:
+        # include_names doesn't work for the fast_reader and RDB (even for unique
+        # colnames), so expect an exception in this case.
+        if rdb is True and fast_reader == 'force':
+            stack.enter_context(pytest.raises(AssertionError))
+
+        dat = ascii.read(lines, fast_reader=fast_reader, include_names=['a', 'a_2', 'a_3'])
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_read.py]: Why is this all under "with ExitStack()"? This seems to be a separate test.

File: astropy/io/ascii/tests/test_read.py [hunk 3]
```diff
@@ -1493,3 +1495,43 @@ def test_kwargs_dict_guess(enable):
...
+        lines += ['N N N N N']
+    lines += ['1 2 3 4 5', '10 20 30 40 50']
+
+    if rdb:
+        lines = ['\t'.join(line.split()) for line in lines]
+
+    dat = ascii.read(lines, fast_reader=fast_reader)
+    assert dat.colnames == ['a', 'a_2', 'a_1', 'a_3', 'a_4']
+
+    with ExitStack() as stack:
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_read.py]: This is so complicated, we need a test tp test that the test tests what we think it does. I suggest splitting this in two tests.

1) ``@pytest.mark.parametrize(scenario, [[False, False], [False, 'force'], [True, False])``
2) a test with no parameters for `rdb=True` and `fast_Reader='force'`

Yes, approximate 8 lines of code will be copied, but it might be easier to read.

Disclaimer: It's entirely possible that this suggestion is inappropriate and I just don't understand what the test is doing.

**Comment by taldcroft** [comment]
@hamogu - I updated the tests (more-or-less) per your suggestion.  It was a bit more than 8 lines but in the end is a little better because it shows xfails now for things that would ideally work, but don't.

@dhomeier - It would be great if you have a chance to look at the fast reader update in the next few days.

**Comment by hamogu** [comment]
I approve (bunt don't click the "approve" button until @dhomeier has looked at the fast reader).

**Comment by dhomeier** [comment]
Sorry this has slipped a bit off the edge of my desk...

The new functionality looks good to me, and I could not see anything directly affecting `tokenizer.c`, but I'll have a closer look over the rest of the weekend.
One thing for clarification: as far as I could see, the failures on the `rdb` format with `fast_reader` are none of the `deduplicate` functionality, but you have rather uncovered another `fast_reader` bug related to the `include_names` option.
On 4.0.1 I am getting
```
lines = ['a\ta_2\ta_1\ta_3\ta_4', 'N\tN\tN\tN\tN', '1\t2\t3\t4\t5', '10\t20\t30\t40\t50']
>>> ascii.read(lines, fast_reader='force')
<Table length=2>
  a    a_2   a_1   a_3   a_4 
int64 int64 int64 int64 int64
----- ----- ----- ----- -----
    1     2     3     4     5
   10    20    30    40    50
>>> ascii.read(lines, fast_reader=True, include_names=['a', 'a_2', 'a_1'])
<Table length=2>
  a    a_2   a_1 
int64 int64 int64
----- ----- -----
    1     2     3
   10    20    30
>>> ascii.read(lines, fast_reader='force', include_names=['a', 'a_2', 'a_1'])
<Table length=3>
 a   a_2  a_1 
str2 str2 str2
---- ---- ----
   N    N    N
   1    2    3
  10   20   30
```
Since your first test in `test_deduplicate_names_basic` already verifies that
```
lines = ['a\ta_2\ta_1\ta\ta', 'N\tN\tN\tN\tN', '1\t2\t3\t4\t5', '10\t20\t30\t40\t50']
>>> ascii.read(rdb, fast_reader='force')
<Table length=2>
  a    a_2   a_1   a_3   a_4 
int64 int64 int64 int64 int64
----- ----- ----- ----- -----
    1     2     3     4     5
   10    20    30    40    50
```
is functioning correctly, I propose in this test to just state what is working and what is not (and maybe rename the rename/modify the `rdb_fast_fail` tests to reflect that they are not testing an exclusive `deduplicate` failure).
So this would be reasonably documented until the actual bug can be fixed in another PR (will try to have a look into that in the next days, too).

File: astropy/io/ascii/tests/test_read.py [hunk 4]
```diff
@@ -1509,3 +1509,69 @@ def test_kwargs_dict_guess(enable):
...
+        lines = ['\t'.join(line.split()) for line in lines]
+    return lines
+
+
+@pytest.mark.parametrize('rdb', [False, True])
+@pytest.mark.parametrize('fast_reader', [False, 'force'])
+def test_deduplicate_names_basic(rdb, fast_reader):
+    """Test that duplicate column names are successfully de-duplicated for the
+    basic format.  Skip the case of rdb=True and fast_reader='force' since that
+    fails and is tested below.
```
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:1530]:
```suggestion
    basic format.  Skip the case of rdb=True and fast_reader='force' when selecting
    include_names, since that fails and is tested below.
```

File: astropy/io/ascii/tests/test_read.py [hunk 5]
```diff
@@ -1509,3 +1509,69 @@ def test_kwargs_dict_guess(enable):
...
+        assert len(dat) == 2
+        assert dat.colnames == ['b1', 'b2', 'b4']
+        assert np.all(dat['b1'] == [1, 10])
+        assert np.all(dat['b2'] == [2, 20])
+        assert np.all(dat['b4'] == [4, 40])
+
+
+@pytest.mark.xfail
+def test_deduplicate_names_rdb_fast_fail1():
+    """Test that duplicate column names fail for the rdb format for fast reader.
```
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:1558]:
```suggestion


@pytest.mark.xfail
def test_include_names_rdb_fast_fail0():
    """Test that selecting column names with `include_names` fails for the rdb format for fast reader.
    This is not desired but reflects a current known limitation.
    """
    lines = _get_lines(True)
    lines[0] = 'a\ta_2\ta_1\ta_3\ta_4'

    dat = ascii.read(lines, fast_reader='force', include_names=['a', 'a_2', 'a_3'])
    assert len(dat) == 2


@pytest.mark.xfail
def test_include_names_rdb_fast_fail1():
    """Test that `include_names` fails with duplicate column names for the rdb format for fast reader.
```

File: astropy/io/ascii/tests/test_read.py [hunk 6]
```diff
@@ -1509,3 +1509,69 @@ def test_kwargs_dict_guess(enable):
...
+    """
+    lines = _get_lines(True)
+
+    dat = ascii.read(lines, fast_reader='force', include_names=['a', 'a_2', 'a_3'])
+    assert len(dat) == 2
+
+
+@pytest.mark.xfail
+def test_deduplicate_names_rdb_fast_fail2():
+    """Test that duplicate column names fail for the rdb format for fast reader.
```
Comment (dhomeier) [inline astropy/io/ascii/tests/test_read.py:1569]:
```suggestion
def test_include_names_rdb_fast_fail2():
    """Test that `include_names` fails with duplicate column names for the rdb format for fast reader.
```

File: astropy/io/ascii/tests/test_read.py (hunk 2 above)
Comment (dhomeier) [reply astropy/io/ascii/tests/test_read.py:1539]: Sorry, only spotted now that `include_names` had already been identified as the culprit in the earlier commits. Still, the current version gives the misleading impression that it is primarily a problem of the duplicate colnames.

**Comment by hamogu** [comment]
For the record: I'm against another option (fix, fix with warning, fail). We have so many options already, and we don't want to complicate things further. 99% of the users will chose whatever the default is anyway, so might just as well only implement that. The remaining 1% can just look at the resulting table by hand, or, if they know how the file should be formatted, test for `np.any(['_1` in c for c in colnames])` or something like that.

**Comment by dhomeier** [comment]
Agree after re-reading the remarks on spurious warnings.

**Comment by dhomeier** [comment]
I may have found a fix for `FastRDB._read_header()` that handles the basic `include`/`exclude_names` failures, but is now messing things up when combined with the custom `names=[...]` option, so clearly needs further analysis.

**Comment by dhomeier** [comment]
@taldcroft some parts of the fix are too distant from this commit's entries to put them in as suggestions, so I've pushed them to my own copy of your branch in 9bf1941. If you prefer them opened in my own PR, let me know what would be the best way to merge this PR into it.

The fix might be a bit winding as I saw no way to directly test inside of `_read_header` for the `*names` kwargs, which are not public in `self.engine` – or is there another way? But this solution introduces one possibly unnecessary read of one header row, which may be tolerable.
Also the last two tests have perhaps become redundant, as they basically test what has already been asserted in `test_deduplicate_names_basic`.

**Comment by taldcroft** [comment]
@dhomeier - thanks! I'll have a look today or tomorrow and see where to go with this.

**Comment by taldcroft** [comment]
@dhomeier - I had a chance to review your proposed fix for the RDB / include_names problem.  I'm happy to see you have enough understanding of the code to get to a fix that works.  Sadly I cannot say the same.  In this case I'm a bit worried that the fix is increasing code complexity noticeably in an already-difficult code and reducing performance a bit while addressing what seems like an exceptional corner case.  ("Perfect is the enemy of good"). 

So I would propose to not merge your fix.  However to ensure your work is not lost, perhaps after this PR is merged (and after the freeze) you could just make that one commit as a PR and then close it. This will give a record of the issue and the fix on the off chance that it comes up again in the wild.  Does that sound agreeable?

**Comment by dhomeier** [comment]
@taldcroft 100 % agree on keeping it out of this PR, as it is a sufficiently separate issue not to block this one any longer.
Nonetheless I tend to keep open the new PR for a while in case some others still wish to tune in.

1. I am not sure the bug is quite such a corner case, as it will pop up whenever one of the standard options for `names`, `include_names`, `exclude_names` is used with RDB tables. Although the results won't change unless `fast_reader` is enforced, which probably does make it less critical (and apparently it has not bothered anyone in real-world use yet).

2. The proposed solution would make 3 calls to `self.engine.read_header()` instead of 2 for parsing the header, so I don't foresee any performance issues – but maybe some systematic testing for really wide tables is in place here.

But all this is indeed best discussed in its own PR, so I'll wait for this one to be merged.

**Comment by taldcroft** [comment]
Thanks @dhomeier.

Does anybody use RDB these days?  The only reason this is included in io.ascii is that there was one person at CfA 15 years ago that was using the whole linux RDB piping stuff (https://www.drdobbs.com/rdb-a-unix-command-line-database/199101326).  So I put it into asciitable because I had to read some of his files.  It seemed nifty enough 20 years ago, but time has flown.

So it seems like an arcane format to me, and then adding in (1) duplicate column names and (2) needing include_names gets it to the level of a corner case that might credibly never happen in the wild.  But honestly I don't know if it is really arcane or not.

Googling on the "rdb file format" turns up a bunch of stuff that is completely different, but I could be missing something.

**Comment by dhomeier** [comment]
> Does anybody use RDB these days? The only reason this is included in io.ascii is that there was one person at CfA 15 years ago that was using the whole linux RDB piping stuff (https://www.drdobbs.com/rdb-a-unix-command-line-database/199101326). So I put it into asciitable because I had to read some of his files. It seemed nifty enough 20 years ago, but time has flown.
> 
That would explain why no issues have popped up so far... ;-)

> So it seems like an arcane format to me, and then adding in (1) duplicate column names and (2) needing include_names gets it to the level of a corner case that might credibly never happen in the wild. But honestly I don't know if it is really arcane or not.

It does not depend in any way on duplicate names, that's what the test I added is checking.
And the underlying problem should actually affect any formats with multiple-line headers, but it seems none of the others have fast reader implementations.

**Comment by hamogu** [comment]
Without a careful look at the code, just from reading the arguments exchanged between the two of you, I'm more and more in favor of including @dhomeier's fix. Yes, it increases code complexity, but @dhomeier has already spend a lot of time in understanding the internals. It's only a moderate extra effort for him to put thorough comments in the code that preserve that knowledge and might even make the new (more complicated) code easier to understand than the current (less thoroughly documented) code. I'm afraid that we rediscover this bug at some later point after we implement another reader with two header lines. By that time, we will have forgotten about this issue and we'll need to do it all again.
